taropsbt/interface.go

```diff
@@ -0,0 +1,141 @@
+)
+
+// VPacket is a PSBT extension packet for a virtual transaction. It represents
+// the virtual asset state transition as it will be validated by the Taro VM.
+// Some elements within the virtual packet may refer to on-chain elements (such
+// as the anchor BTC transaction that was used to anchor the input that is
+// spent). But in general a virtual transaction does NOT directly map onto a BTC
+// transaction. It is entirely possible that multiple virtual transactions will
+// be merged into a single BTC transaction. Thus, each asset state transfer is
+// represented in a virtual TX and multiple asset state transfers can be
+// anchored within a single BTC transaction.
+//
+// TODO(guggero): Actually support merging multiple virtual transactions into a
+// single BTC transaction.
+type VPacket struct {
+	// Input is the asset input that is being spent.
+	//
+	// TODO(guggero): Support spending multiple inputs.
+	Input *VInput
+
+	// Outputs is the list of new asset outputs that are created by the
+	// virtual transaction. By convention the output at index 0 is the
+	// change output for asset change or the asset tombstone in case of a
+	// non-interactive full value send.
+	Outputs []*VOutput
+}
+
+// VInput represents an input to a virtual asset state transition transaction.
+type VInput struct {
+	// PInput is the embedded default PSBT input struct that is used for
+	// already existing data such as the Taproot internal key or script
+	// witness.
+	psbt.PInput
+
+	// PrevID is the asset previous ID of the asset being spent.
+	PrevID asset.PrevID
+
+	// AnchorPoint is the outpoint that the Commitment below is anchored on
+	// in the main chain. When serialized, this will be stored as the
+	// PreviousOutPoint of the wire.TxIn of the PSBT's unsigned TX.
+	AnchorPoint wire.OutPoint
+
+	// AnchorOutputValue is output value of the anchor output.
+	AnchorOutputValue btcutil.Amount
+
+	// AnchorOutputInternalKey is the internal key of the anchor output that
+	// this input is spending the asset from.
+	AnchorOutputInternalKey *btcec.PublicKey
+
+	// AnchorOutputBip32Derivation is the BIP32 derivation of the anchor
+	// output's internal key.
+	//
+	// TODO(guggero): Do we also want to allow multiple derivations to be
+	// specified here? That would allow us to specify multiple keys involved
+	// in MuSig2 for example. Same for the Taproot derivation below.
+	AnchorOutputBip32Derivation *psbt.Bip32Derivation
+
+	// AnchorOutputTaprootBip32Derivation is the Taproot BIP32 derivation of
+	// the anchor output's internal key.
+	AnchorOutputTaprootBip32Derivation *psbt.TaprootBip32Derivation
+
+	// TapscriptSibling is the tapscript sibling of this asset. This will
+	// usually be blank.
+	TapscriptSibling []byte
+
+	// Asset is the full instance of the asset being spent.
+	Asset *asset.Asset
+
+	// IsSplit indicates whether the input asset is being split into more
+	// than one piece. This is always true for non-interactive sends.
+	IsSplit bool
```

Comment:
I think this should also have an `mssmt.Proof` object embedded? 

Depending on the assumptions/use, it can either have the above, or even just a full `proof.AssetSnapshot`. 

If both sides have already verified the inputs, then this isn't needed. However, if both sides are collaborating without that, then provide the last asset proof for the input either lets them just accept that, or use a Universe to further query backwards for verification.

Reply:
This is mostly just a convenience flag. This being false means it's a full value interactive send. So I can also extract that into a helper function so we don't need to encode it in the packet. Yeah, maybe that's a good idea.